fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java

```diff
@@ -748,7 +771,223 @@ private void validateDocsInPartition(Index index, int partitionId, Tuple groupin
+        final int partitionCount;
+        if (repartitionCount == 3) {
+            partitionCount = allIds.size() / partitionSize + 1;
+            assertThat(segmentCounts, Matchers.aMapWithSize(partitionCount));
+            assertEquals(IntStream.range(0, partitionCount).boxed()
+                            .collect(Collectors.toMap(Function.identity(), partitionId -> partitionId == partitionCount - 1 ? 1 : 2)),
+                    segmentCounts);
+        } else {
+            partitionCount = allIds.size() / partitionSize;
+            assertThat(segmentCounts, Matchers.aMapWithSize(partitionCount));
+            assertEquals(IntStream.range(0, partitionCount).boxed()
+                            .collect(Collectors.toMap(Function.identity(), partitionId -> 2)),
+                    segmentCounts);
+        }
+
+        try (FDBRecordContext context = openContext(contextProps)) {
+            schemaSetup.accept(context);
+            validateDocsInPartition(index, 0, groupingKey, repartitionCount == 3 ? 8 : 10,
+                    allIds.stream()
+                            .skip(repartitionCount == 3 ? 192 : 190)
+                            .map(idLong -> Tuple.from(docGroupFieldValue, idLong))
+                            .collect(Collectors.toSet()),
+                    "text:propose");
+            for (int i = 1; i < 20; i++) {
+                // 0 should have the newest
+                // everyone else should increase
+                validateDocsInPartition(index, i, groupingKey, partitionSize,
+                        allIds.stream().skip((i - 1) * partitionSize)
+                                .limit(partitionSize)
+                                .map(idLong -> Tuple.from(docGroupFieldValue, idLong))
+                                .collect(Collectors.toSet()),
+                        "text:propose");
+            }
+            List<LucenePartitionInfoProto.LucenePartitionInfo> partitionInfos = getPartitionMeta(index,
+                    groupingKey, contextProps, schemaSetup);
+            assertEquals(partitionCount, partitionInfos.size());
+        }
+    }
+
+    static Stream<Arguments> randomizedRepartitionTest() {
+        // This has found situations that should have explicit tests:
+        //      1. Multiple groups
+        //      2. When the size of first partition is exactly highWatermark+repartitionCount
+        return Stream.concat(
+                // at the time of writing, 13 isn't special, but it gives us one seed, the other two found issues
+                Stream.of(
+                        Arguments.of(13L),
+                        Arguments.of(-644766138635622644L),
+                        Arguments.of(-1089113174774589435L)),
+                RandomizedTestUtils.randomArguments(random -> Arguments.of(random.nextLong())));
+    }
+
+    @ParameterizedTest
+    @MethodSource
+    void randomizedRepartitionTest(long seed) throws IOException {
```

Comment:
I've run this a few times with 100 iterations with randomness, all successful.
I only added a couple here to be run regularly.

It might make sense to have this have more arguments (e.g. `isGrouped`/`isSynthetic`/`groupCount`) and then have a seed for all the other stuff so that we can more easily guarantee that the non-random test has various combinations. I may tackle that if I feel I have the cyles, or defer for later.

Reply:
The changes in #2495 caused this test to break sometimes because there are more arrangements that can occur. I pushed a new change that is more precise in the assertions, and have run it 1000 times successfully.